pkg/scalers/gcp_stackdriver_scaler.go

```diff
@@ -0,0 +1,190 @@
+			gcpStackdriverLog.Error(err, "Error parsing targetValue")
+			return nil, fmt.Errorf("error parsing targetValue: %s", err.Error())
+		}
+
+		meta.targetValue = targetValue
+	}
+
+	auth, err := getGcpAuthorization(config, config.ResolvedEnv)
+	if err != nil {
+		return nil, err
+	}
+	meta.gcpAuthorization = auth
+	return &meta, nil
+}
+
+func (s *stackdriverScaler) IsActive(ctx context.Context) (bool, error) {
+	value, err := s.getMetrics(ctx)
+	if err != nil {
+		gcpStackdriverLog.Error(err, "error getting metric value")
+		return false, err
+	}
+	return value > 0, nil
+}
+
+func (s *stackdriverScaler) Close(context.Context) error {
+	if s.client != nil {
+		err := s.client.metricsClient.Close()
+		s.client = nil
+		if err != nil {
+			gcpStackdriverLog.Error(err, "error closing StackDriver client")
+		}
+	}
+
+	return nil
+}
+
+// GetMetricSpecForScaling returns the metric spec for the HPA
+func (s *stackdriverScaler) GetMetricSpecForScaling(context.Context) []v2beta2.MetricSpec {
+	// Construct the target value as a quantity
+	targetValueQty := resource.NewQuantity(int64(s.metadata.targetValue), resource.DecimalSI)
+
+	externalMetric := &v2beta2.ExternalMetricSource{
+		Metric: v2beta2.MetricIdentifier{
+			Name: s.metadata.metricName,
+		},
+		Target: v2beta2.MetricTarget{
+			Type:         v2beta2.AverageValueMetricType,
+			AverageValue: targetValueQty,
+		},
+	}
+
+	// Create the metric spec for the HPA
+	metricSpec := v2beta2.MetricSpec{
+		External: externalMetric,
+		Type:     externalMetricType,
+	}
+
+	return []v2beta2.MetricSpec{metricSpec}
+}
+
+// GetMetrics connects to Stack Driver and retrieves the metric
+func (s *stackdriverScaler) GetMetrics(ctx context.Context, metricName string, metricSelector labels.Selector) ([]external_metrics.ExternalMetricValue, error) {
+	value, err := s.getMetrics(ctx)
+	if err != nil {
+		gcpStackdriverLog.Error(err, "error getting metric value")
+		return []external_metrics.ExternalMetricValue{}, err
+	}
+
+	metric := external_metrics.ExternalMetricValue{
+		MetricName: metricName,
+		Value:      *resource.NewQuantity(value, resource.DecimalSI),
+		Timestamp:  metav1.Now(),
+	}
+
+	return append([]external_metrics.ExternalMetricValue{}, metric), nil
+}
+
+func (s *stackdriverScaler) setStackdriverClient(ctx context.Context) error {
+	var client *StackDriverClient
+	var err error
+	if s.metadata.gcpAuthorization.podIdentityProviderEnabled {
+		client, err = NewStackDriverClientPodIdentity(ctx)
+	} else {
+		client, err = NewStackDriverClient(ctx, s.metadata.gcpAuthorization.GoogleApplicationCredentials)
+	}
+
+	if err != nil {
+		gcpStackdriverLog.Error(err, "Failed to create stack driver client")
+		return err
+	}
+	s.client = client
+	return nil
+}
+
+// getMetrics gets metric type value from stackdriver api
+func (s *stackdriverScaler) getMetrics(ctx context.Context) (int64, error) {
+	if s.client == nil {
+		err := s.setStackdriverClient(ctx)
+		if err != nil {
+			return 0, err
+		}
+	}
```

Comment:
Could it be possible to initialize this client during the constructor? Does it make sense?

Reply:
I guess this way, if the client fails to initialize for some reason (permissions, transient comm. issue) it will be established the next time.

Reply:
In case of error inside the scaler (any ), KEDA will refresh the internal cache and will recreate the whole scaler, so that behavior is already achieved for [operator](https://github.com/kedacore/keda/blob/main/pkg/scaling/cache/scalers_cache.go#L172-L190) and also [metrics server](https://github.com/kedacore/keda/blob/main/pkg/scaling/cache/scalers_cache.go#L65-L80)